jlm/rvsdg/traverser.cpp

```diff
@@ -193,258 +222,275 @@ TopDownTraverser::onInputChange(Input * in, Output * old_output, Output * new_ou
 {
-  markVisited(node);
-}
+  const auto node = TryGetOwnerNode<Node>(*input->origin());
+  if (!node)
+    return;
 
-void
-BottomUpTraverser::onNodeDestroy(Node * node)
-{
-  for (const auto & input : node->Inputs())
+  if (isInputActivated(*input))
   {
-    if (auto pred = TryGetOwnerNode<Node>(*input.origin()))
-    {
-      // Set threshold to 0 here: The predecessor node is
-      // still connected, so its successor count is not correct.
-      // However, if the node has been activated before, then
-      // it will remain activated after this removal. The only
-      // thing that we need to ensure here is that the total
-      // count is correct.
-      tracker_.decActivationCount(pred, 0);
-    }
+    tracker_.incActivationCount(node, node->numSuccessors());
+  }
+  else
+  {
+    tracker_.checkNodeDeactivation(node, node->numSuccessors());
   }
 }
 
+template<bool IsConst>
 void
-BottomUpTraverser::onInputCreate(Input * input)
+BottomUpTraverserGeneric<IsConst>::onInputChange(
+    Input * input,
+    Output * oldOrigin,
+    Output * newOrigin)
 {
-  if (auto pred = TryGetOwnerNode<Node>(*input->origin()))
+  const auto inputActive = isInputActivated(*input);
+
+  if (auto oldNode = TryGetOwnerNode<Node>(*oldOrigin))
   {
-    if (isInputActivated(input))
+    if (inputActive)
     {
-      tracker_.incActivationCount(pred, pred->numSuccessors());
+      tracker_.decActivationCount(oldNode, oldNode->numSuccessors());
     }
     else
     {
-      tracker_.checkNodeDeactivation(pred, pred->numSuccessors());
+      // The oldNode might have just lost its final non-active user
+      tracker_.checkNodeActivation(oldNode, oldNode->numSuccessors());
     }
   }
-}
 
-void
-BottomUpTraverser::onInputChange(Input * in, Output * old_origin, Output * new_origin)
-{
-  if (isInputActivated(in))
+  if (auto newNode = TryGetOwnerNode<Node>(*newOrigin))
   {
-    if (auto pred = TryGetOwnerNode<Node>(*old_origin))
+    if (inputActive)
     {
-      tracker_.decActivationCount(pred, pred->numSuccessors());
+      tracker_.incActivationCount(newNode, newNode->numSuccessors());
     }
-    if (auto pred = TryGetOwnerNode<Node>(*new_origin))
+    else
     {
-      tracker_.incActivationCount(pred, pred->numSuccessors());
+      tracker_.checkNodeDeactivation(newNode, newNode->numSuccessors());
     }
   }
 }
 
+template<bool IsConst>
 void
-BottomUpTraverser::onInputDestroy(Input * input)
+BottomUpTraverserGeneric<IsConst>::onInputDestroy(Input * input)
 {
-  if (auto pred = TryGetOwnerNode<Node>(*input->origin()))
+  const auto node = TryGetOwnerNode<Node>(*input->origin());
+  if (!node)
+    return;
+
+  if (isInputActivated(*input))
   {
-    if (isInputActivated(input))
-    {
-      tracker_.decActivationCount(pred, 0);
-    }
-    else
-    {
-      tracker_.checkNodeActivation(pred, pred->numSuccessors() - 1);
-    }
+    tracker_.decActivationCount(node, node->numSuccessors() - 1);
```

Comment:
This line changed compared to the previous code.

Reply:
This change makes no difference, so I changed this back to 0 and added a comment to explain.

The bugfix I was thinking about was in `onInputChange`, where it would do nothing if the input was not activated. However, moving a not activated input might remove the last non-activated user of a node, which should place the node on the frontier. Likewise it may cause the new origin node to no longer be on the frontier

Reply:
Yes  good point, it should have been separate